src/foam/parse/PropertyParser.java

```diff
@@ -0,0 +1,145 @@
+/**
+ * @license
+ * Copyright 2018 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package foam.parse;
+
+import java.util.List;
+
+import foam.core.PropertyInfo;
+import foam.lib.json.KeyParser;
+import foam.lib.json.Whitespace;
+import foam.lib.parse.Alt;
+import foam.lib.parse.Literal;
+import foam.lib.parse.PStream;
+import foam.lib.parse.ParserContext;
+import foam.lib.parse.ProxyParser;
+import foam.lib.parse.Seq3;
+import foam.mlang.MLang;
+import foam.mlang.predicate.Predicate;
+
+public class PropertyParser extends ProxyParser {
+  private PropertyInfo property;
+
+  public PropertyParser(PropertyInfo p) {
+    super(new Seq3(4, 3,
+        new Whitespace(),
+        new Alt(
+          new KeyParser(
+            p.getName()),
+          new KeyArrayParser(
+            p.getAliases()),
+          new KeyParser(
+            p.getShortName())),
+        new Whitespace(),
+        new Alt(
+          new Literal(":"),
+          new Literal("="),
+          new Literal("!="),
+          new Literal("<="),
+          new Literal(">="),
+          new Literal("<"),
+          new Literal(">"),
+          new Literal("=>"),
+          new Literal("=<")),
+          p.queryParser(),
+        new Whitespace()));
+    property = p;
+  }
+
+  public PStream parse(PStream ps, ParserContext x) {
+    Predicate p = null;
+    ps = super.parse(ps, x);
+    if (ps == null)
+      return null;
+
+    if (!(ps.value() instanceof Object[])) {
+      property.set(x.get("obj"), ps.value());
+      p = createPredicate(ps.operator().toString(), x);
+    } else {// Just for date
+      p = createDatePredicate(ps.operator().toString(), ps.value());
+    }
+
+    if (p != null) {
+      // the NOT predicate is apply on property
+      if (((List<Predicate>) x.get("objPar")).size() > 0) {
+        Predicate lastPredicate = ((List<Predicate>) x.get("objPar"))
+            .get(((List<Predicate>) x.get("objPar")).size() - 1);
+        if (lastPredicate instanceof foam.mlang.predicate.Not) {
+          ((List<Predicate>) x.get("objPar")).remove(((List<Predicate>) x.get("objPar")).size() - 1);
+          ((List<Predicate>) x.get("objPar")).add(MLang.NOT(p));
+        }
+      }
+      ((List<Predicate>) x.get("objPar")).add(p);
+    }
+    return ps;
+  }
+
+  protected Predicate createDatePredicate(String operator, Object dates) {
+    Predicate p = null;
+    Object date1 = ((Object[]) dates)[0];
+    Object date2 = ((Object[]) dates)[1];
+
+    switch (operator) {
+    case ":":
+    case "=":

```

Comment:
See how in my example we don't need the addition of "operator()" to the PStream?

Instead the operator() is defined by the particular parser that matches, the LtExpression parser just returns an actual foam.mlang.predicate.Lt(), so we don't need to add these switch statements.

Can we move this towards my style?